src/core/components/pin.js

```diff
@@ -0,0 +1,438 @@
+          if (err) {
+            return cb(err)
+          }
+
+          nodes.forEach(node => {
+            const key = toB58String(node.multihash)
+            if (!recursivePins.has(key)) {
+              // not already pinned recursively or directly
+              indirectKeys.add(key)
+            }
+          })
+
+          cb()
+        })
+      }, (err) => {
+        if (err) { return callback(err) }
+        callback(null, Array.from(indirectKeys))
+      })
+    }),
+
+    // encodes and writes pin key sets to the datastore
+    // each key set will be stored as a DAG node, and a root node will link to both
+    flush: promisify((callback) => {
+      const newInternalPins = new Set()
+      const logInternalKey = (mh) => newInternalPins.add(toB58String(mh))
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+      waterfall([
+        // create link to direct keys node
+        (cb) => pin.set.storeSet(pin.directKeys(), logInternalKey, cb),
+        (dRoot, cb) => DAGLink.create(pin.types.direct, dRoot.size, dRoot.multihash, cb),
+        (dLink, cb) => handle.put('dLink', dLink, cb),
+        // create link to recursive keys node
+        (cb) => pin.set.storeSet(pin.recursiveKeys(), logInternalKey, cb),
+        (rRoot, cb) => DAGLink.create(pin.types.recursive, rRoot.size, rRoot.multihash, cb),
+        (rLink, cb) => handle.put('rLink', rLink, cb),
+        // the pin-set nodes link to an empty node, so make sure it's added to dag
+        (cb) => DAGNode.create(Buffer.alloc(0), cb),
+        (empty, cb) => dag.put(empty, {cid: new CID(empty.multihash)}, cb),
+        // create root node with links to direct and recursive nodes
+        (cid, cb) => DAGNode.create(Buffer.alloc(0), [handle.dLink, handle.rLink], cb),
+        (root, cb) => handle.put('root', root, cb),
+        // add the root node to dag
+        (cb) => dag.put(handle.root, {cid: new CID(handle.root.multihash)}, cb),
+        // update the internal pin set
+        (cid, cb) => cb(null, logInternalKey(handle.root.multihash)),
+        // save serialized root to datastore under a consistent key
+        (_, cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null), // hack for CLI tests
+        (_, cb) => repo.datastore.put(pinDataStoreKey, handle.root.multihash, cb)
+      ], (err, result) => {
+        if (err) { return callback(err) }
+        self.log(`Flushed pins with root: ${handle.root}.`)
+        internalPins = newInternalPins
+        return callback(null, handle.root)
+      })
+    }),
+
+    load: promisify((callback) => {
+      const newInternalPins = new Set()
+      const logInternalKey = (mh) => newInternalPins.add(toB58String(mh))
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+      waterfall([
+        (cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null), // hack for CLI tests
```

Comment:
Has the case where no `~/.jsipfs` directory exists been handled?  I can't see anything obvious in the PR code. 
 `repo.datastore` will not exist until the user runs `jsipfs init`.

Reply:
Oh good catch. If there is no ~/.jsipfs, it's impossible to do any datastore based operations such as `jsipfs object new`. We should have a common pattern for this, like `assertRepoExists` where appropriate. I think this falls under the umbrella of: https://github.com/ipfs/js-ipfs/issues/1180.

Code written just for tests is always something to try to avoid, though, and I hadn't looked at this line again so I will now.